src/envoy/http/jwt_auth/jwt.cc

```diff
@@ -421,27 +485,67 @@ void Pubkeys::CreateFromJwksCore(const std::string &pkey_jwks) {
   }
 
   for (auto jwk_json : keys) {
-    std::unique_ptr<Pubkey> pubkey(new Pubkey());
-
-    std::string n_str, e_str;
     try {
-      pubkey->kid_ = jwk_json->getString("kid");
-      pubkey->alg_ = jwk_json->getString("alg");
-      pubkey->alg_specified_ = true;
-      n_str = jwk_json->getString("n");
-      e_str = jwk_json->getString("e");
+      ExtractPubkeyFromJwk(jwk_json);
     } catch (Json::Exception &e) {
       continue;
     }
-    EvpPkeyGetter e;
-    pubkey->key_ = e.EvpPkeyFromJwk(n_str, e_str);
-    keys_.push_back(std::move(pubkey));
   }
+
   if (keys_.size() == 0) {
     UpdateStatus(Status::JWK_NO_VALID_PUBKEY);
   }
 }
 
+void Pubkeys::ExtractPubkeyFromJwk(Json::ObjectSharedPtr jwk_json) {
+  // Check "kty" parameter, it should exist.
+  // https://tools.ietf.org/html/rfc7517#section-4.1
+  // If "kty" is missing, throw an exception.
+  std::string kty = jwk_json->getString("kty", "");
+  if (kty != "RSA" && kty != "EC") {
+    return;
+  }
+  // Compare "kty" and "alg", and do nothing if they do not match.
+  // https://tools.ietf.org/html/rfc7518#section-6.1
+  // https://tools.ietf.org/html/rfc7518#section-7.1.2
+  std::string alg = jwk_json->getString("alg");
+  if (kty == "EC" && alg == "ES256") {
+    ExtractPubkeyFromJwkEC(jwk_json);
+  } else if (kty == "RSA" && alg.compare(0, 2, "RS") == 0) {
+    ExtractPubkeyFromJwkRSA(jwk_json);
+  }
+}
+
+void Pubkeys::ExtractPubkeyFromJwkRSA(Json::ObjectSharedPtr jwk_json) {
+  std::unique_ptr<Pubkey> pubkey(new Pubkey());
+  pubkey->kid_ = jwk_json->getString("kid");
+  pubkey->alg_ = jwk_json->getString("alg");
```

Comment:
need to try and catch to handle invalid format

Reply:
Done. Thanks.